[feature] docker build tooling

sescobb27 opened the pull request:
** Usage

```ex
path = "/Users/kiro/workspace/elixir/elixir-docker-guide"
{ctx, _} = 
  Path.join([path, "Dockerfile"])
  |> DockerApi.DockerfileParser.parse()
  |> DockerApi.DockerBuild.build(path)
```

sescobb27 committed:
[add] initial commit for docker build integration via remote API

sescobb27 committed:
[add] initial implementation of the Dockerfile parser in elixir

sescobb27 committed:
[refactor] improve the Dockerfile parser

crsuarez commented on apps/docker_api/lib/dockerfile_parser.ex:
```diff
@@ -0,0 +1,90 @@
+    acc
+  end
+
+  # first line - accomulator empty
```
accomulator -> accumulator

sescobb27 committed:
[change] do not use File.cd!/2 because is not multi-process safe

sescobb27 committed:
[remove] unused test file

sescobb27 committed:
[change] improve docker api wrapper
[add] more abstractions to the docker api

sescobb27 committed:
[WIP][add] initial version of the docker build feature

sescobb27 committed:
[add] support for FROM image with or without as alias

sescobb27 committed:
[add] inject path to expand on instructions

[add] support for exec and shell form instructions

sescobb27 committed:
[add] support new commands in the docker build and fix issues

[add] support for `COPY`ing files from host machine to container
[add] `COPY`ing files doesn't support relative PATH to WORKDIR it has to be
and absolute path
[fix] take only ImageId parte when committing a container
[add] add some logs to commands
[add] support 204 and 304 when stopping a container
[add] limit kill container after 5 sec waiting for a container
[add] remove generated TAR file and return the ContainerId
[add] wait :infinity when pulling an image
[fix] capture only the matching part of the regexes
[add] wait for some instructions to finish
[add] support for ENTRYPOINT command

sescobb27 committed:
[add] inject custom wait times (true, false, or any number)

[refactor] DockerApi to be much cleaner

sescobb27 committed:
[WIP][ci skip] trying to implement mounting a volum at build time, still not working

sescobb27 committed:
[add] merge 2 context maps

sescobb27 committed:
[add] docker file example for testing bind mounts

sescobb27 committed:
[add] bind mount host file system at container build time

[refactor] inject context to every command instead of just the ImageId
[refactor] stop container before commiting it

sescobb27 committed:
[change] move all HTTP calls into a different module to separate concerns

sescobb27 committed:
[refactor] raname parse function name
[fix] dialyzer errors
[change] how regexes capture matches

sescobb27 committed:
[add] dialyzer specs but still failing due to interaction with erlang BIF
[change] handle erl_tar errors

sescobb27 committed:
[change] refactor interaction with the Docker Remote API
[change] handle API errors
[change] use railway oriented flow for creating a layer

sescobb27 committed:
[change] handle docker api errors
[change] fail fast in the presence of errors
[change] use rail way oriented flow when uploading dir/files to a container
[add] typespecs

sescobb27 committed:
[change] update readme with awesome introduction
